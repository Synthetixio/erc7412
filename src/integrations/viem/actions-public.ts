import * as viem from 'viem'

import {
  prepareTransactionRequest as actionPrepareTransactionRequest,
  estimateContractGas as actionEstimateContractGas,
  simulateContract as actionSimulateContract
} from 'viem/actions'

import { callWithOffchainData } from '../..'
import type { OracleAdapter } from '../../types'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412PublicActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      call: async (args: viem.CallParameters): Promise<viem.CallReturnType> => {
        return {
          data: (
            await callWithOffchainData(
              [
                {
                  from: (args.account as any)?.address ?? viem.zeroAddress,
                  ...args
                }
              ],
              client,
              adapters
            )
          )[0]
        }
      },
      readContract: async (args: viem.ReadContractParameters): Promise<viem.ReadContractReturnType> => {
        return {
          data: viem.decodeFunctionResult({
            ...args,
            data: (
              await callWithOffchainData(
                [
                  {
                    from: (args.account as any)?.address ?? viem.zeroAddress,
                    data: viem.encodeFunctionData(args),
                    ...args
                  }
                ],
                client,
                adapters
              )
            )[0]
          })
        }
      },
      prepareTransactionRequest: async (args: viem.PrepareTransactionRequestParameters) => {
        try {
          return await actionPrepareTransactionRequest(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for prepareTransactionRequest')
          throw err
        }
      },
      estimateContractGas: async (args: viem.EstimateContractGasParameters) => {
        try {
          return await actionEstimateContractGas(client, args)
        } catch (err) {
          console.log('WARN: erc7412 not implemented for estimateContractGas')
          throw err
        }
      },
      // TODO: types
      simulateContract: async (args: viem.SimulateContractParameters): Promise<any> => {
        try {
          return await actionSimulateContract(client, args)
        } catch (err) {
          const baseTxn = {
            from: (args.account as any)?.address ?? viem.zeroAddress,
            to: args.address,
            chain: args.chain,
            data: viem.encodeFunctionData(args),
            value: args.value
          }
          return {
            result: viem.decodeFunctionResult({
              ...args,
              data: (await callWithOffchainData([baseTxn], client, adapters))[0]
            }) as any
          }
        }
      },
      multicall: async (args: viem.MulticallParameters): Promise<viem.MulticallReturnType> => {
        if (args.contracts.length < 1) {
          throw new Error('must have at least one call for multicall')
        }

        const retvals = await callWithOffchainData(
          // todo: types have a problem with the fact it cannot verify that the array is at least 1 long
          args.contracts.map((c) => {
            return {
              from: c.address ?? viem.zeroAddress,
              data: viem.encodeFunctionData(c),
              ...c
            }
          }) as any,
          client,
          adapters
        )

        return retvals.map((r, i) => {
          return { result: viem.decodeFunctionResult({ ...args.contracts[i], data: r }), status: 'success' }
        })
      }
      // below functions are not included because they are not applicable
      // writeContract: async (args: viem.WriteContractParameters) => {},
    }

    return actions
  }
}
