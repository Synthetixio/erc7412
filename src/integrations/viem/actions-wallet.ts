import * as viem from 'viem'

import { prepareTransactionRequest as actionPrepareTransactionRequest } from 'viem/actions'

import { resolvePrependTransaction, makeTrustedForwarderMulticall, callWithOffchainData } from '../..'

import ITrustedMulticallForwarder from '../../../out/ITrustedMulticallForwarder.sol/ITrustedMulticallForwarder.json'

import type { TransactionRequest } from '../..'
import type { OracleAdapter } from '../../types'
import { getAccount } from './actions-public'

/**
 * Extend your viem client with the object returned by this function to automatically apply erc7412
 * required offchain data to your read calls
 */
export function createErc7412WalletActions (adapters: OracleAdapter[]) {
  return (client: viem.PublicClient) => {
    const actions = {
      prepareTransactionRequest: async (
        args: viem.PrepareTransactionRequestParameters
      ): Promise<viem.PrepareTransactionRequestReturnType> => {
        return await actions.prepareMulticallTransactionRequest({ txns: [args] })
      },

      prepareMulticallTransactionRequest: async (args: {
        txns: viem.PrepareTransactionRequestParameters[]
      }): Promise<viem.PrepareTransactionRequestReturnType> => {
        const prependedTxns: TransactionRequest[] = []

        if (args.txns.length < 1) {
          throw new Error('must have at least 1 transaction in multicall')
        }

        const payloadTxns: TransactionRequest[] = args.txns.map((t) => {
          const req: TransactionRequest = {
            ...t,
            from: getAccount(t.account)
          }

          return req
        })

        while (true) {
          const multicallTxn: viem.PrepareTransactionRequestParameters =
            payloadTxns.length > 1 || prependedTxns.length > 0
              ? {
                  ...makeTrustedForwarderMulticall([...prependedTxns, ...payloadTxns]),
                  account: args.txns[0].account,
                  chain: args.txns[0].chain
                }
              : args.txns[0]
          try {
            return await actionPrepareTransactionRequest(client, multicallTxn)
          } catch (err) {
            prependedTxns.push(await resolvePrependTransaction(err, client, adapters))
          }
        }
      },
      simulateContract: async (args: viem.SimulateContractParameters): Promise<viem.SimulateContractReturnType> => {
        const baseTxn = {
          from: (args.account as any)?.address ?? viem.zeroAddress,
          to: args.address,
          chain: args.chain,
          data: viem.encodeFunctionData(args),
          value: args.value
        }
        const preparedTxn = await actions.prepareTransactionRequest(baseTxn)
        return {
          request:
            preparedTxn.to === args.address
              ? args
              : {
                  ...preparedTxn,
                  abi: ITrustedMulticallForwarder.abi,
                  address: preparedTxn.to,
                  functionName: 'aggregate3Value',
                  args: viem.decodeFunctionData({ abi: ITrustedMulticallForwarder.abi, data: preparedTxn.data ?? '0x' })
                },
          result: viem.decodeFunctionResult({
            ...args,
            data: (await callWithOffchainData([baseTxn], client, adapters))[0]
          })
        } as any // TODO
      }

      // NOTE: cant override `sendRawTransaction` because the transaction is
      // already serialized and so its impossible to impart a new multicall
      // in front without breaking permissions
      // `sendTransaction` is not included because its behavior is not supposed to affect the transaction,
      // its just supposed to sign and send
    }

    return actions
  }
}
