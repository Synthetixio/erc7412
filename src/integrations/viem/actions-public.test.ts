import * as viem from 'viem'
import type { PublicClient } from 'viem'
import type { OracleAdapter } from '../../types'

import { createErc7412PublicActions } from './actions-public'

import { callWithOffchainData } from '../..'

jest.mock('../..')

const fakeAddress = viem.getContractAddress({ from: viem.zeroAddress, nonce: 0n })

export const fakeAdapters: OracleAdapter[] = [
  {
    getOracleId: () => 'FAKE',
    fetchOffchainData: async () => [{ arg: '0x87651234' as viem.Hex, fee: BigInt(100) }]
  }
]

describe('createErc7412PublicActions', () => {
  const mockPublicClient = {
    getBlockNumber: jest.fn().mockResolvedValue(12345),
    simulateContract: jest.fn().mockResolvedValue({
      /* Mock result */
    })
    // ...other methods as needed
  } as unknown as PublicClient

  const actions = createErc7412PublicActions(fakeAdapters)(mockPublicClient)

  const fakeTxn = {
    to: fakeAddress,
    data: '0x5678',
    value: 203n
  } as const

  const fakeContractRequest = {
    abi: [
      {
        type: 'function',
        name: 'greet',
        inputs: [
          {
            type: 'string',
            name: 'first'
          },
          {
            type: 'string',
            name: 'second'
          }
        ],
        outputs: [
          {
            type: 'uint32',
            name: 'result'
          }
        ],
        stateMutability: 'view'
      }
    ] as viem.Abi,
    address: fakeAddress,
    functionName: 'greet',
    args: ['uno', 'dos']
  }

  describe('call()', () => {
    it('should pass through a single transaction', async () => {
      jest.mocked(callWithOffchainData).mockResolvedValue(['0x83838'])
      const result = await actions.call(fakeTxn)
      expect(result).toEqual({ data: '0x83838' })
      expect(jest.mocked(callWithOffchainData).mock.calls[0][0]).toEqual([{ from: viem.zeroAddress, ...fakeTxn }])
    })
  })

  describe('readContract()', () => {
    it('should pass through a single transaction', async () => {
      jest
        .mocked(callWithOffchainData)
        .mockResolvedValue(['0x0000000000000000000000000000000000000000000000000000000000000009'])
      const result = await actions.readContract(fakeContractRequest)
      expect(result).toEqual({ data: 9 })
      expect(jest.mocked(callWithOffchainData).mock.calls[0][0][0]).toMatchObject({
        data: viem.encodeFunctionData(fakeContractRequest)
      })
    })
  })
})
