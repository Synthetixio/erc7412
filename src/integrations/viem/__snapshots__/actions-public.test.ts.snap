// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integrations/viem/actions-public createErc7412PublicActions() multicall() errors if nothing to multicall 1`] = `"must have at least one call for multicall"`;
